libmscore/unrollrepeats.cpp

```diff
@@ -0,0 +1,352 @@
+      QList<Element*> elems;
+      score->scanElements(&elems,collectElements,false);
+      for (auto e : elems)
+            if (e->isMarker() || e->isJump()) {
+                  score->deleteItem(e);
+                  }
+            else if (e->isBarLine()) {
+                  ((BarLine*)e)->setBarLineType(BarLineType::NORMAL);
+                  }
+
+      // Set the last bar line to end symbol
+      score->lastMeasure()->setEndBarLineType(BarLineType::END, false);
+      Segment * last = score->lastMeasure()->segments().last();
+      if (last->segmentType() == SegmentType::EndBarLine) {
+            auto els = last->elist();
+            for (uint i=0;i<els.size();i++) {
+                  if (!els[i]) continue;
+                  ((BarLine*)(els[i]))->setBarLineType(BarLineType::END);
+                  }
+            }
+      }
+
+
+//---------------------------------------------------------
+//   createExcerpts
+//    re-create all the excerpts once the master score
+//    has been unrolled
+//---------------------------------------------------------
+
+void createExcerpts(MasterScore * cs, QList<Excerpt *> excerpts)
+      {
+      // From musescore.cpp endsWith(".pdf")
+      for (Excerpt * e: excerpts) {
+            Score * nscore = new Score(e->oscore());
+            e->setPartScore(nscore);
+            nscore->style().set(Sid::createMultiMeasureRests, true);
+            cs->startCmd();
+            cs->undo(new AddExcerpt(e));
+            Excerpt::createExcerpt(e);
+
+            // Borrowed from excerptsdialog.cpp
+            // a new excerpt is created in AddExcerpt, make sure the parts are filed
+            for (Excerpt* ee : e->oscore()->excerpts()) {
+                  if (ee->partScore() == nscore && ee != e) {
+                        ee->parts().clear();
+                        ee->parts().append(e->parts());
+                        }
+                  }
+
+            cs->endCmd();
+            }
+
+      qDebug() << "Created new excerpts:" << cs->excerpts().size();
+      }
+
+//---------------------------------------------------------
+//   unrollRepeats
+//    unroll all the repeats
+//---------------------------------------------------------
+
+MasterScore * MasterScore::unrollRepeats()
+      {
+
+      MasterScore * original = this;
+      
+      // Create a copy of the original score to play with
+      MasterScore * score = original->clone(); 
+
+      // Figure out repeat structure
+      original->repeatList()->unwind();
+      original->setPlaylistDirty();
+
+      //If no repeats, just return the score as-is
+      if (original->repeatList()->count()==1)
+            return score;
+
+      // Remove excerpts for now (they are re-created after unrolling master score)
+      QList<Excerpt *> excerpts;
+      for (Excerpt * e : ((MasterScore*)score)->excerpts()) {
+            excerpts.append(e);
+            qDebug() << "Removing ", e->title();
+            score->masterScore()->deleteExcerpt(e);
```

Comment:
Don't really get how that would work. The excerpts need to be re-generated but copying them would not achieve that. Or would it?

Reply:
No, copying itself wouldn't, but you could add that copy to `excerpts` list and use it as you do it right now. I just point that using the original `Excerpt` object for that purpose may be potentially dangerous.

Reply:
@dmitrio95 I am honestly admitting I do not understand the partScores and excerpts well enough to know if what I am doing is correct or not, but I changed
excerpts.append(e) to excerpts.append(e->clone());
which, if I understood you correctly, should take care of this issue for now.
Please do let me know if that is not quite what you meant, or if you believe other steps need to be taken.

Reply:
@dmitrio95  ... or I would have if it had a copy constructor or a clone function, but it does not...
So for now, I replaced it with
`            Excerpt * ne = new Excerpt(score);
            ne->setTitle(e->title());
            ne->setParts(e->parts());`
Which seems to be sufficient for the current purposes.
I could go in and add a copy constructor to Excerpt too, but in this case, more fields likely need to be copied for it to work in other use cases, and since I really do not understand how that part of the code works, it would be rather dangerous for me to just leave it there...
@dmitrio95 @anatoly-os how would you recommend I proceed?

Reply:
Copying tracks mapping (`e->tracks()`) would also be necessary to restore correctly those excerpts which have separate voices extracted rather than whole parts. Other than that, using a copying constructor would certainly seem to be a better solution but defining a proper copy constructor for `Excerpt` would require copying a `partScore` which is redundant here (and may be really slow). Maybe a compromise could be defining a copy constructor with some boolean `copyPartScore` parameter, something like:
```
Excerpt(const Excerpt& ex, bool copyPartScore = true)
   : _oscore(ex._oscore), _title(ex._title), _parts(ex._parts), _tracks(ex._tracks)
      {
      _partScore = (copyPartScore && ex._partScore) ? ex._partScore->clone() : nullptr;
      }
```
Then it would be possible both not to worry about `Excerpt`'s internals in this code and avoid the unnecessary score copying.

Reply:
@dmitrio95 added the constructor and used it in my code with copyPartScore set to false as you recommended.